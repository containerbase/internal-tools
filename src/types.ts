export enum Commands {
  BinaryBuilder = 'binary-builder',
  DockerConfig = 'docker-config',
  DockerBuilder = 'docker-builder',
}
