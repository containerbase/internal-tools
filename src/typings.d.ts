/* eslint-disable  */
declare module 'www-authenticate' {
  namespace wwwAuthenticate {
    namespace parsers {
      class WWW_Authenticate {
        readonly parms: Record<string, string>;
        constructor(auth: string);
      }
    }
  }
  export = wwwAuthenticate;
}

declare namespace NodeJS {
  interface Global {
    renovateCache: unknown;
    repoCache: unknown;
  }
}
