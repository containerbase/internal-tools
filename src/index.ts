// Node v16 is missing structuredClone, so we polyfill it here.
import 'core-js/actual/structured-clone';
import run from './runner';

run().catch(console.error);
