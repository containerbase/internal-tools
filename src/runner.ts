import { getInput, setFailed } from '@actions/core';
import * as chalk from 'chalk';
import { Commands } from './types';
import log from './utils/logger';

export default async function run(): Promise<void> {
  try {
    log.info(chalk.blue('Renovate Docker Builder'));
    const cmd = getInput('command') as Commands;
    log.info(chalk.yellow('Executing:'), ` ${cmd}`);
    switch (cmd) {
      case Commands.BinaryBuilder:
        await (await import('./commands/binary')).run();
        break;
      case Commands.DockerBuilder:
        await (await import('./commands/docker/builder')).run();
        break;

      case Commands.DockerConfig:
        await (await import('./commands/docker/config')).run();
        break;

      default:
        log.error(chalk.red('Unknown command:'), cmd);
        break;
    }
  } catch (error) {
    setFailed((error as Error).message);
  }
}
