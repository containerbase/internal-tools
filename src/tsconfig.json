{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "rootDir": ".",
    "outDir": "dist",
    "types": ["node"]
  }
}
