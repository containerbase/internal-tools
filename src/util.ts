import { exec as _exec } from '@actions/exec';
import { ExecOptions as _ExecOptions } from '@actions/exec/lib/interfaces';
import log from './utils/logger';
import { getInput, startGroup, endGroup } from '@actions/core';
import { join } from 'path';
import fs from 'fs';
import { promisify } from 'util';
import findUp = require('find-up');

const readFileAsync = promisify(fs.readFile);

export type ExecOptions = _ExecOptions;

export type ExecResult = {
  readonly code: number;
  readonly stdout: string;
  readonly stderr: string;
};

export async function exec(
  cmd: string,
  args: string[],
  options?: ExecOptions
): Promise<ExecResult> {
  let stdout = '';
  let stderr = '';
  let code: number;

  try {
    startGroup(`${cmd} ${args.join(' ')}`);
    code = await _exec(cmd, args, {
      ...options,
      listeners: {
        stdout: (data: Buffer) => {
          stdout += data.toString();
        },
        stderr: (data: Buffer) => {
          stderr += data.toString();
        },
      },
    });
  } finally {
    endGroup();
  }
  if (code) {
    log.error({ code });
    throw new Error('Command failed');
  }

  return { code, stdout, stderr };
}

export function isDryRun(): boolean {
  const val = getInput('dry-run');
  return !!val && val === 'true';
}

/**
 * Get environment variable or empty string.
 * Used for easy mocking.
 * @param key variable name
 */
export function getEnv(key: string): string {
  return process.env[key] ?? '';
}

export async function readJson<T = unknown>(file: string): Promise<T> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  const res = await import(path);
  // istanbul ignore next
  return res?.default ?? res;
}

export async function readFile(file: string): Promise<string> {
  const path = join(getEnv('GITHUB_WORKSPACE'), file);
  return await readFileAsync(path, 'utf8');
}

export const MultiArgsSplitRe = /\s*(?:[;,]|$)\s*/;

export function getArg(name: string, opts?: { required?: boolean }): string;
export function getArg(
  name: string,
  opts?: { required?: boolean; multi: true }
): string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[];
export function getArg(
  name: string,
  opts?: { required?: boolean; multi?: boolean }
): string | string[] {
  const val = getInput(name, opts);
  return opts?.multi ? val.split(MultiArgsSplitRe).filter(Boolean) : val;
}

let _pkg: Promise<string | undefined>;

/**
 * Resolve path for a file relative to renovate root directory (our package.json)
 * @param file a file to resolve
 */
export async function resolveFile(file: string): Promise<string> {
  if (!_pkg) {
    _pkg = findUp('package.json', { cwd: __dirname, type: 'file' });
  }
  const pkg = await _pkg;
  // istanbul ignore if
  if (!pkg) {
    throw new Error('Missing package.json');
  }
  return join(pkg, '../', file);
}
