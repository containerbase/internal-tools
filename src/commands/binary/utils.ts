import { getInput } from '@actions/core';
import is from '@sindresorhus/is';
import { getDefaultVersioning } from 'renovate/dist/modules/datasource';
import { getArch, getArg, getDistro, isDryRun, readJson } from '../../util';
import { readDockerConfig } from '../../utils/config';
import {
  DockerPlatform,
  dockerBuildx,
  dockerRun,
} from '../../utils/docker/common';
import log from '../../utils/logger';
import type { BinaryBuilderConfig, ConfigFile } from '../../utils/types';

export async function getConfig(): Promise<BinaryBuilderConfig> {
  const configFile = getInput('config') || 'builder.json';

  const cfg = await readJson<ConfigFile>(configFile);

  if (!is.object(cfg)) {
    throw new Error('missing-config');
  }

  if (!is.string(cfg.image)) {
    cfg.image = getInput('image', { required: true });
  }

  if (!is.string(cfg.buildArg)) {
    cfg.buildArg = cfg.image.toUpperCase() + '_VERSION';
  }

  await readDockerConfig(cfg);

  return {
    ...cfg,
    ignoredVersions: cfg.ignoredVersions ?? [],
    dryRun: isDryRun(),
    lastOnly: getArg('last-only') == 'true',
    buildArgs: getArg('build-args', { multi: true }),
    versioning: cfg.versioning ?? getDefaultVersioning(cfg.datasource),
  } as BinaryBuilderConfig;
}

export async function createBuilderImage(
  ws: string,
  { buildArgs }: BinaryBuilderConfig
): Promise<void> {
  log('Creating builder image');
  const args = [
    'build',
    '--load',
    '-t',
    'builder',
    '--build-arg',
    `DISTRO=${getDistro()}`,
  ];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  if (is.nonEmptyArray(buildArgs)) {
    args.push(...buildArgs.map((b) => `--build-arg=${b}`));
  }
  await dockerBuildx(...args, ws);
}

export async function runBuilder(ws: string, version: string): Promise<void> {
  const args = ['--name', 'builder', '--volume', `${ws}/.cache:/cache`];
  const arch = getArch();
  if (is.nonEmptyString(arch)) {
    args.push('--platform', DockerPlatform[arch]);
  }
  await dockerRun(...args, 'builder', version);
}
