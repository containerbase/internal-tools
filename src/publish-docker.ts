import { getInput } from '@actions/core';
import { exec } from './util';
import chalk from 'chalk';
import log from 'fancy-log';
import { getRemoteImageId, getLocalImageId } from './utils/docker';

export const MultiArgsSplitRe = /\s*(?:;|$)\s*/;
export async function run(): Promise<void> {
  const dryRun = !!getInput('dry-run');
  const image = getInput('image');
  let tags = getInput('tags')
    ?.split(MultiArgsSplitRe)
    .filter(Boolean);

  if (!tags?.length) {
    tags = ['latest'];
  }

  if (!image) {
    throw new Error('Missing image');
  }

  for (const tag of tags) {
    const fullName = `${image}:${tag}`;
    log.info(chalk.blue('Processing image:'), chalk.yellow(fullName));

    log('Fetch new id');
    const newId = await getLocalImageId(image, tag);

    log('Fetch old id');
    const oldId = await getRemoteImageId(image, tag);

    if (oldId === newId) {
      log('Image uptodate, no push nessessary: ', chalk.yellow(oldId));
      continue;
    }

    log('Publish new image', `${oldId} > ${newId}`);
    if (dryRun) {
      log.warn(chalk.yellow('DRY-RUN: Would push: '), fullName);
    } else {
      await exec('docker', ['push', fullName]);
    }
    log.info(chalk.blue('Processing image finished: ', newId));
  }
}
