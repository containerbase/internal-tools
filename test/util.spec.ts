import { existsSync } from 'fs';
import * as _core from '@actions/core';
import * as _exec from '@actions/exec';
import * as _io from '@actions/io';
import * as util from '../src/util';
import { mocked } from './utils';

jest.mock('@actions/core');
jest.mock('@actions/exec');
jest.mock('@actions/io');
jest.mock('../src/utils/logger');

const core = mocked(_core);
const exec = mocked(_exec);
const io = mocked(_io);

describe('util', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('exists', () => {
    it('returns true', async () => {
      io.which.mockResolvedValueOnce('');
      expect(await util.exists('test')).toBe(true);
    });

    it('returns false', async () => {
      io.which.mockRejectedValueOnce(new Error());
      expect(await util.exists('test')).toBe(false);
    });
  });

  describe('exec', () => {
    it('works', async () => {
      exec.exec.mockImplementationOnce((_cmd, _args, opts) => {
        opts?.listeners?.stdout && opts.listeners.stdout(Buffer.from('test'));
        opts?.listeners?.stderr &&
          opts.listeners.stderr(Buffer.from('testerr'));
        return Promise.resolve(0);
      });
      expect(await util.exec('dummy-cmd', [])).toEqual({
        code: 0,
        stdout: 'test',
        stderr: 'testerr',
      });
    });

    it('throws', async () => {
      expect.assertions(1);
      exec.exec.mockResolvedValueOnce(1);
      await expect(util.exec('dummy-cmd', [])).rejects.toThrow();
    });
  });

  describe('isDryRun', () => {
    afterEach(() => {
      delete process.env.CI;
    });

    it('false', () => {
      process.env.CI = 'true';
      core.getInput.mockReturnValueOnce('false');
      expect(util.isDryRun()).toBe(false);
      expect(util.isDryRun()).toBe(false);
      expect(core.getInput).toHaveBeenCalledTimes(2);
    });

    it('true', () => {
      core.getInput.mockReturnValueOnce('true');
      expect(util.isDryRun()).toBe(true);
      expect(core.getInput).toHaveBeenCalledTimes(1);
    });
  });

  describe('getEnv', () => {
    it('works', () => {
      expect(util.getEnv('NOT_FOUND_ENV_VAR')).toBe('');
      expect(util.getEnv('PATH')).toBeDefined();
    });
  });

  describe('getWorkspace', () => {
    let ws: string | undefined;

    beforeEach(() => {
      ws = process.env.GITHUB_WORKSPACE;
      delete process.env.GITHUB_WORKSPACE;
    });

    afterEach(() => {
      process.env.GITHUB_WORKSPACE = ws;
    });

    it('works', () => {
      expect(util.getWorkspace()).toBe(process.cwd());
      process.env.GITHUB_WORKSPACE = '/var/test';
      expect(util.getWorkspace()).toBe('/var/test');
    });
  });

  it('getDistro', () => {
    expect(util.getDistro()).toBe('focal');
  });

  it('getArch', () => {
    expect(util.getArch()).toBe('');
  });

  describe('readJson', () => {
    afterEach(() => {
      delete process.env.GITHUB_WORKSPACE;
    });

    it('works', async () => {
      process.env.GITHUB_WORKSPACE = process.cwd();
      expect(await util.readJson('.prettierrc.json')).toEqual({
        singleQuote: true,
        trailingComma: 'es5',
      });
    });
  });

  describe('readFile', () => {
    afterEach(() => {
      delete process.env.GITHUB_WORKSPACE;
    });

    it('works', async () => {
      process.env.GITHUB_WORKSPACE = process.cwd();
      expect(await util.readFile('Dockerfile')).toMatchSnapshot();
    });
  });

  describe('readBuffer', () => {
    afterEach(() => {
      delete process.env.GITHUB_WORKSPACE;
    });

    it('works', async () => {
      process.env.GITHUB_WORKSPACE = process.cwd();
      expect(
        (await util.readBuffer('Dockerfile')).toString()
      ).toMatchSnapshot();
    });
  });

  describe('getArg', () => {
    it('single', () => {
      core.getInput.mockReturnValueOnce('test;latest;slim');
      expect(util.getArg('dockerfile')).toBe('test;latest;slim');
    });

    it('multi', () => {
      core.getInput.mockReturnValueOnce('test;latest;slim');
      expect(util.getArg('dockerfile', { multi: true })).toEqual([
        'test',
        'latest',
        'slim',
      ]);
    });

    it('multi (null)', () => {
      expect(util.getArg('dockerfile', { multi: true })).toEqual([]);
    });
  });

  describe('resolveFile', () => {
    it('works', async () => {
      const file = await util.resolveFile('package.json');
      expect(file).toBeDefined();
      expect(existsSync(file)).toBe(true);

      const file2 = await util.resolveFile('bin/dummy.sh');
      expect(file2).toBeDefined();
      expect(existsSync(file2)).toBe(false);
    });
  });
});
