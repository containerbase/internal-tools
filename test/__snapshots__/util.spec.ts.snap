// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` readFile works 1`] = `
"FROM renovate/renovate:19.225.3-slim

FROM scratch

# renovate: datasource=npm depName=yarn versioning=npm
ARG YARN_VERSION=1.22.0

# renovate: datasource=npm depName=pnpm versioning=npm
ARG PNPM_VERSION=4.0.0

# renovate: datasource=docker depName=python versioning=docker
ARG PYTHON_VERSION=3.7.7

# renovate: datasource=github-releases depName=helm/helm
ARG HELM_VERSION=3.4.0

"
`;
